src/event_list_proxy.h

```diff
@@ -0,0 +1,56 @@
+/// @file
+/// SPDX-License-Identifier: GPL-3.0-or-later
+/// @author Simon Heybrock
+/// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+/// National Laboratory, and European Spallation Source ERIC.
+#ifndef EVENT_LIST_PROXY_H
+#define EVENT_LIST_PROXY_H
+
+#include "range/v3/view/zip.hpp"
+
+#include "dataset.h"
+#include "zip_view.h"
+
+// TODO Rename to ConstEventListProxy and add EventListProxy, inheriting from
+// ConstEventListProxy.
+class EventListProxy {
+public:
+  // TODO Fix ZipView to work with const Dataset, or use something else here.
+  EventListProxy(Dataset &dataset) : m_dataset(&dataset) {}
+  EventListProxy(const typename Data::EventTofs_t::type &tofs,
+                 const typename Data::EventPulseTimes_t::type &pulseTimes)
+      : m_tofs(&tofs), m_pulseTimes(&pulseTimes) {
+    if (tofs.size() != pulseTimes.size())
+      throw std::runtime_error("Size mismatch for fields of event list.");
+  }
+
+  // TODO Either ZipView must be generalized, or we need to use a different view
+  // type here, once we support another event storage format. Furthermore, we
+  // want to support read-only access if only a subset of all fields is
+  // requested, e.g., for reading only TOF, without need to know whether also
+  // pulse-times or weights are present.
+  template <class... Tags>
+  ZipView<Tags...> getMutable(const Tags... tags) const {
+    return ZipView<Tags...>(*m_dataset);
+  }
+
+  // TODO This should be template/overloaded for different field combination,
+  // e.g., Tof-only, with weights, ...
+  auto get() const {
+    if (m_dataset) {
+      const Dataset *d(m_dataset);
+      return ranges::view::zip(d->get(Data::Tof), d->get(Data::PulseTime));
+    }
+    return ranges::view::zip(
+        gsl::make_span(m_tofs->data(), m_tofs->data() + m_tofs->size()),
+        gsl::make_span(m_pulseTimes->data(),
+                       m_pulseTimes->data() + m_tofs->size()));
+  }
+
+private:
+  Dataset *m_dataset{nullptr};
+  const typename Data::EventTofs_t::type *m_tofs{nullptr};
+  const typename Data::EventPulseTimes_t::type *m_pulseTimes{nullptr};
+};
```

Comment:
Slightly orthogonal to this, but with the re-design would it be useful to think about how event compression could be nicely supported. i.e. for monitors where storing raw events might be too expensive, but grouping events by pulse would also allow us to support filtering data from monitors against sample environment.

This would not happen at the level of the proxy.

Reply:
On the contrary, I think *everything* happens at the level of the proxy. You can put anything in the dataset, the aggregation into an event-list look-alike is done by the proxy.

I don't think event-compression changes anything, it simply needs to be implemented in the proxy once needed.

Reply:
The proxy is pointing to existing tofs_times, pulse_times. To avoid the memory overhead you may want to apply the compression as early as possible (before forming event lists to begin with).

Reply:
dataset would allow us to store an additional uncertainty for the pulse times in the case that compression was applied. I don't think this is possible to do in Mantid at present.

Reply:
> The proxy is pointing to existing tofs_times, pulse_times. To avoid the memory overhead you may want to apply the compression as early as possible (before forming event lists to begin with).

Still, this is nothing that `Dataset` needs to be aware of, it is just a dumb container. Either you put the data directly into the correct variable (or combination of variables), or you use a proxy. Performing the compression is just an implementation of an algorithm and I do not see how that would influence the design.